casr/src/bin/casr-ubsan.rs

```diff
@@ -0,0 +1,450 @@
+    // Update warning vector
+    // Get position by input
+    let mut warnings: Vec<(UbsanWarning, CrashLine)> = vec![];
+    for warning in extracted_warnings {
+        // Get crashline
+        if let Ok(crashline) = warning.crash_line() {
+            warnings.push((warning, crashline));
+        } else {
+            warn!(
+                "Cannot get warning crashline {:?} from {:?}",
+                warning, input
+            );
+        }
+    }
+
+    Ok(warnings)
+}
+
+/// Generate ubsan report for specified input file
+///
+/// # Arguments
+///
+/// * `input` - input file path
+///
+/// * `warning` - target warning
+///
+/// * `crashline` - warning crashile
+///
+/// * `argv` - target program argument vector
+///
+/// * `pre_report` - report template containing identic values
+///
+/// # Returns value
+///
+/// Generated report
+///
+fn gen_report(
+    input: &Path,
+    warning: &UbsanWarning,
+    crashline: &CrashLine,
+    argv: &[&str],
+    pre_report: &CrashReport,
+) -> CrashReport {
+    // Get command line argv
+    let mut argv = argv.to_owned();
+    let arg: String;
+    let stdin = if let Some(index) = argv.iter().position(|&arg| arg.contains("@@")) {
+        arg = argv[index].replace("@@", input.to_str().unwrap());
+        argv[index] = &arg;
+        false
+    } else {
+        true
+    };
+    let args = argv.join(" ");
+    debug!("Generating reports for {:?}", args);
+    // Create report
+    let mut report = pre_report.clone();
+    report.proc_cmdline = args;
+    report.ubsan_report = warning.ubsan_report();
+    if stdin {
+        report.stdin = input.to_str().unwrap().to_string();
+    }
+    // Get stacktrace
+    if let Ok(stacktrace) = warning.extract_stacktrace() {
+        report.stacktrace = stacktrace;
+    }
+    // Get execution class
+    if let Ok(execution_class) = warning.severity() {
+        report.execution_class = execution_class;
+    }
+    // Get crashline and source
+    report.crashline = crashline.to_string();
+    if let CrashLine::Source(debug) = crashline {
+        if let Some(sources) = CrashReport::sources(debug) {
+            report.source = sources;
+        }
+    }
+    report
+}
+
+/// Save ubsan report
+///
+/// # Arguments
+///
+/// * `report` - saving report
+///
+/// * `output_dir` - report saving directory
+///
+/// * `input` - input file path (need for unique name)
+///
+fn save_report(report: CrashReport, output_dir: &Path, input: &Path) -> Result<()> {
+    // Convert report to string.
+    let repstr = serde_json::to_string_pretty(&report).unwrap();
+
+    let dir_name = input.parent().unwrap().file_name().unwrap();
+    let input_name = input.file_name().unwrap();
+    let crashline = report.crashline;
+    let mut crashline = crashline.split('/').last().unwrap();
+    // Drop column number
+    let re = Regex::new(r#".+:\d+:\d"#).unwrap();
+    if re.is_match(crashline) {
+        let re = Regex::new(r#"(.+:\d+):\d"#).unwrap();
```

Comment:
Could you use this regex instead of first one?

Reply:
or simply call split(':')

Reply:
Done

Reply:
Done x2